servant-server/src/Servant/Server/Internal.hs

```diff
@@ -60,10 +63,11 @@ import           Servant.Server.Internal.Router
 import           Servant.Server.Internal.RoutingApplication
 import           Servant.Server.Internal.ServantErr
 
+
 class HasServer layout where
   type ServerT layout (m :: * -> *) :: *
 
-  route :: Proxy layout -> IO (RouteResult (Server layout)) -> Router
+  route :: Proxy layout -> Delayed (Server layout) -> Router
```

Comment:
It's not quite as obvious how to do `IO` in the combinator instances (as for example in [auth-combinator.hs](https://github.com/haskell-servant/servant/blob/master/servant-examples/auth-combinator/auth-combinator.hs#L36)). I imagine we'd need a `MonadIO` instance (which to my mind is another reason to consider switching to `StateT`).